README.md
# x65

6502 Macro Assembler in a single c++ file using the struse single file text parsing library. Supports most syntaxes. x65 was recently named Asm6502 but was renamed because Asm6502 is too generic, x65 has no particular meaning.

![x65](bin/x65.png)

In order to minimize the documentation and make this page shorter I've moved the [old documentation](../../wiki/Previous-first-page) here.

The [up to date documentation is here](x65.txt).

x65 can assemble 6502, 65C02 and 65816 source and build executables for c64, Apple II or just raw binary.

Noteworthy features:

* Code with sections, object files and linking or single file fixed
  address, or mix it up with fixed address sections in object files.
* Assembler listing with cycle counting for code review.
* Export multiple binaries with a single link operation.
* C style scoping within '{' and '}' with local and pool labels
  respecting scopes.
* Conditional assembly with if/ifdef/else etc.
* Recursible macro support
* Assembler directives representing a variety of features.
* Local labels can be defined in a number of ways, such as leading
  period (.label) or leading at-sign (@label) or terminating
  dollar sign (label$), and leading colon for merlin (:label).
* String Symbols system allows building user expressions and macros
  during assembly.
* Reassignment of symbols and labels by default.
* No indentation required for instructions, meaning that labels can't
  be mnemonics, macros or directives (merlin requires indentation).
* Supporting the syntax of other 6502 assemblers (Merlin syntax
  requires command line argument, -endm adds support for sources
  using macro/endmacro and repeat/endrepeat combos rather
  than scoeps).
* Apple II GS executable output (relocatable executable).

## Features

* **Code**
* **Linking**
* **Comments**
* **Labels**
* **String Symbols**
* **Directives**
* **Macros**
* **Expressions**
* **List File with Cycle Count**

## Prerequisite

x65.cpp requires struse.h which is a single file text parsing library that can be retrieved from https://github.com/Sakrac/struse.

### References

* [6502 opcodes](http://www.6502.org/tutorials/6502opcodes.html)
* [6502 cheat sheet](https://drive.google.com/file/d/0B4lBG-q6aGsROEtKWXRXVnlTNUE/view)
* [6502 opcode grid](http://www.llx.com/~nparker/a2/opcodes.html)
* [Codebase64 CPU section](http://codebase64.org/doku.php?id=base:6502_6510_coding)
* [6502 illegal opcodes](http://www.oxyron.de/html/opcodes02.html)
* [65816 opcodes](http://wiki.superfamicom.org/snes/show/65816+Reference#fn:14)

### Download Binaries

* [Windows x64 binaries](../..//raw/master/bin/x65_x64.zip)
* [Windows x86 binaries](../..//raw/master/bin/x65_win32.zip)

### x65

x65 is the assembler

### x65macro.i

x65macro.i is a 6502 include file that defines a number of standard macros that can assign values, move values, copy values and loop constructs, see x65.txt for details.

### dump_x65

dump_x65 is a tool to inspect the contents of .x65 object files generated by x65 to track down linking issues

### x65dsasm

x65dsasm is a tool to disassemble assembled binary code for review, it will perform a basic analysis and assign labels where appropriate and treats unreferenced bytes as data rather than code. It can also export assemblable code from a binary.

### Acknowledgments

This project would not be completed without the direct or indirect support of great people, some which I can currently remember:

* Marc dePeo, helping me uncover the strange and unique world of Merlin's assembler syntax (and working together with me on True Crime NY gameplay code and more)
* Che Lalic, explaining the murky bits of 65816 (and a Ninja on SNES NBA Hangtime and other projects)
* John Brooks, sharing the Rastan Apple II source code so I could test 65816 and figure out a number of issues with my initial linker, and encouraging the implementation of Apple II GS OS executable file format / OMF export (and helping out with Playstation All-Stars)
* [Brutal Deluxe](http://www.brutaldeluxe.fr) for releasing the excellent OMF Analyzer tool and the source, which was a significant help generating Apple II GS OS executables.
* The C64 demo scene for sharing a great deal of 6502 programming resources and overall inspiration.
* Jordan Mechner, sharing the Prince of Persia Apple II source code so I could test out a significant part of the assembler and the Merlin syntax mode
* Bill Budge, sharing the Pinball Construction Set Apple II source code, although at the point I tried it, all of it just assembled without any assembler issues at all.

### Development Status

Looking for help testing various features of the assembler, I have a large number of tests that pass without fail but there are so many ways for assemblers to break.
Primarily tested with personal archive of sources written for Kick assmebler, DASM, TASM, XASM, etc. and passing most of Apple II Prince of Persia and Pinball Construction set.

**TODO**
* irp (indefinite repeat)

**FIXED**
* Adding MERGE directive, Label Pools rewrite, TEXT data can be indexed from a string symbol
* Label Pools were destroyed after each scope so they did not work in include files which defeated their purpose. Label pools are now persistent through scopes.
* Labels reserved from label pools now distinguish between global and local. Use [.!@$] as a prefix to reserve a local label from a label pool (previously always local)
* Merlin macro parameters are not required on the MAC line, scope braces ('{', '}') can be used in the first column in Merlin.
* First line of a Merlin macro was sometimes ignored, two sequential subtractions were ignored in expressions.
* Pushing source contexts (macro, rept, include etc.) will always increment the scope depth.
* Fixed REPT / LUP to not destroy local symbols in the scope it was used in while also destroying local symbols within the repeating block correctly
* Switched over to inttypes.h from built-in types since the word unsigned was used a little too much in the code
* Removed the disassembler and put it into its own project [x65dsasm](http://github.com/sakrac/x65dsasm)
* LUP/REPT directives clean up local symbols each iteration to avoid crossing over an iteration with branches to local labels.
* Fixed Merlin MAC directive which is a little different from normal assembler macros
* Labels can start with numbers and values will only be interpreted as decimal numbers if terminated by a character that is not an alphabetic character or underscore
* INCSYM failed with local labels, this is now properly handled. (fixed again..)
* INCBIN and IMPORT BINARY always failed (force 0 bytes length)
* Using more than 16 bytes of Pool labels was flawed as byte 15 and 16 would be the same address.
* Fixed STRUCT directive (failed if contained line was empty).
* Adding x65macro.i
* Vice symbols will generate breakpoints whenever label 'debugbreak' is encountered
* Evaluating '==' was broken
* Macros can have dots in their names
* Handling double negative in expressions (--35 == 35)
* Macros works within conditionals (if/else/endif, etc)
* String symbols broke late evaluation resulting in garbage references, this has been fixed
* Added string symbols
* Resolved the DirectPage_Stack section vs. Zeropage section for Apple II GS/OS executables.
* OMF export for Apple II GS/OS executables

[(older fixes)](../../wiki/fixes)

Revisions:
* 10 - String Symbols
* 9 - Apple II GS OS executable
* 8 - Fish food / Linking tested and passed with external project (Apple II gs Rastan)
* 7 - 65816 Support
* 6 - 65C02 support
* 5 - Merlin syntax
* 4 - Object files, relative sections and linking
* 3 - 6502 full support
* 2 - Moved file out of struse samples
* 1 - Built a sample of a simple assembler within struse


vscode/CHANGELOG.md
## 0.1.1
* Converted XML to JSON as XML is the official file format of unreadability
* Added windows install batch file

## 0.1.0
* Bunched together a shabby working version of Visual Studio Code editing of 6502 intended for assembling with x65
* Added coloring for section control

dump_x65/dump_x65.cpp
//
//  dump_x65.cpp
//  dump_x65
//
//  Created by Carl-Henrik Skårstedt on 10/26/15.
//  Copyright © 2015 Carl-Henrik Skårstedt. All rights reserved.
//
// The MIT License (MIT)
//
// Copyright (c) 2015 Carl-Henrik Skårstedt
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software
// and associated documentation files (the "Software"), to deal in the Software without restriction,
// including without limitation the rights to use, copy, modify, merge, publish, distribute,
// sublicense, and/or sell copies of the Software, and to permit persons to whom the Software
// is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or
// substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
// PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
// FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
// Details, source and documentation at https://github.com/Sakrac/x65.
//
// "struse.h" can be found at https://github.com/Sakrac/struse, only the header file is required.
//

#define _CRT_SECURE_NO_WARNINGS		// Windows shenanigans
#define STRUSE_IMPLEMENTATION		// include implementation of struse in this file

#include <stdlib.h>
#include <stdio.h>
#include <inttypes.h>
#include "struse.h"

//
//
// OBJECT FILE HANDLING
//
//

enum SectionType : char {
	ST_UNDEFINED,			// not set
	ST_CODE,				// default type
	ST_DATA,				// data section (matters for GS/OS OMF)
	ST_BSS,					// uninitialized data section
	ST_ZEROPAGE				// ununitialized data section in zero page / direct page
};

struct ObjFileHeader {
	int16_t id;	// 'x6'
	int16_t sections;
	int16_t relocs;
	int16_t labels;
	int16_t late_evals;
	int16_t map_symbols;
	uint32_t stringdata;
	int bindata;
};

struct ObjFileStr {
	int offs;				// offset into string table
};

struct ObjFileSection {
	enum SectionFlags {
		OFS_DUMMY,
		OFS_FIXED,
		OFS_MERGED,
	};
	struct ObjFileStr name;
	struct ObjFileStr exp_app;
	int start_address;
	int end_address;
	int output_size;		// assembled binary size
	int align_address;
	int16_t next_group;		// next section of group
	int16_t first_group;		// first section of group
	int16_t relocs;
	SectionType type;
	int8_t flags;
};

struct ObjFileReloc {
	int base_value;
	int section_offset;
	int16_t target_section;
	int8_t bytes;
	int8_t shift;
};

struct ObjFileLabel {
	enum LabelFlags {
		OFL_EVAL = (1<<15),			// Evaluated (may still be relative)
		OFL_ADDR = (1<<14),			// Address or Assign
		OFL_CNST = (1<<13),			// Constant
		OFL_XDEF = OFL_CNST-1		// External (index into file array)
	};
	struct ObjFileStr name;
	int value;
	int flags;				// 1<<(LabelFlags)
	int16_t section;			// -1 if resolved, file section # if section rel
	int16_t mapIndex;			// -1 if resolved, index into map if relative
};

struct ObjFileLateEval {
	struct ObjFileStr label;
	struct ObjFileStr expression;
	int address;			// PC relative to section or fixed
	int target;				// offset into section memory
	int16_t section;			// section to target
	int16_t rept;				// value of rept for this late eval
	int16_t scope;			// PC start of scope
	int16_t type;				// label, byte, branch, word (LateEval::Type)
};

struct ObjFileMapSymbol {
	struct ObjFileStr name;	// symbol name
	int value;
	int16_t section;
	bool local;				// local labels are probably needed
};

enum ShowFlags {
	SHOW_SECTIONS = 1,
	SHOW_RELOCS = 2,
	SHOW_LABELS = 4,
	SHOW_MAP_SYMBOLS = 8,
	SHOW_LATE_EVAL = 16,
	SHOW_CODE_RANGE = 32,

	SHOW_DEFAULT = SHOW_SECTIONS
};

char* LoadBinary(const char* filename, size_t &size)
{
	if (FILE *f = fopen(filename, "rb")) {
		fseek(f, 0, SEEK_END);
		size_t _size = ftell(f);
		fseek(f, 0, SEEK_SET);
		if (char *buf = (char*)malloc(_size)) {
			fread(buf, _size, 1, f);
			fclose(f);
			size = _size;
			return buf;
		}
		fclose(f);
	}
	size = 0;
	return nullptr;
}

strref PoolStr(struct ObjFileStr off, const char *str_pool)
{
	return off.offs>=0 ? strref(str_pool + off.offs) : strref();
}

enum LEType {			// When an expression is evaluated late, determine how to encode the result
	LET_LABEL,			// this evaluation applies to a label and not memory
	LET_ABS_REF,		// calculate an absolute address and store at 0, +1
	LET_ABS_L_REF,		// calculate a bank + absolute address and store at 0, +1, +2
	LET_ABS_4_REF,		// calculate a 32 bit number
	LET_BRANCH,			// calculate a branch offset and store at this address
	LET_BRANCH_16,		// calculate a branch offset of 16 bits and store at this address
	LET_BYTE,			// calculate a byte and store at this address
};

static const char *late_type[] = {
	"LABEL", "ABS_REF", "ABS_L_REF", "ABS_4_REF",
	"BRANCH", "BRANCH_16", "BYTE" };
static const char *section_type[] = {
	"UNDEFINED",		// not set
	"CODE",				// default type
	"DATA",				// data section (matters for GS/OS OMF)
	"BSS",				// uninitialized data section
	"ZEROPAGE"			// ununitialized data section in zero page / direct page
};
static const int section_type_str = sizeof(section_type) / sizeof(section_type[0]);


void ReadObjectFile(const char *file, uint32_t show = SHOW_DEFAULT)
{
	size_t size;
	if (char *data = LoadBinary(file, size)) {
		struct ObjFileHeader &hdr = *(struct ObjFileHeader*)data;
		size_t sum = sizeof(hdr) + hdr.sections*sizeof(struct ObjFileSection) +
		hdr.relocs * sizeof(struct ObjFileReloc) + hdr.labels * sizeof(struct ObjFileLabel) +
		hdr.late_evals * sizeof(struct ObjFileLateEval) +
		hdr.map_symbols * sizeof(struct ObjFileMapSymbol) + hdr.stringdata + hdr.bindata;
		if (hdr.id == 0x7836 && sum == size) {
			struct ObjFileSection *aSect = (struct ObjFileSection*)(&hdr + 1);
			struct ObjFileReloc *aReloc = (struct ObjFileReloc*)(aSect + hdr.sections);
			struct ObjFileLabel *aLabels = (struct ObjFileLabel*)(aReloc + hdr.relocs);
			struct ObjFileLateEval *aLateEval = (struct ObjFileLateEval*)(aLabels + hdr.labels);
			struct ObjFileMapSymbol *aMapSyms = (struct ObjFileMapSymbol*)(aLateEval + hdr.late_evals);
			const char *str_orig = (const char*)(aMapSyms + hdr.map_symbols);
			size_t code_start = sum - hdr.bindata, code_curr = code_start;

			// sections
			if (show & SHOW_SECTIONS) {
				int reloc_idx = 0;
				for (int si = 0; si < hdr.sections; si++) {
					struct ObjFileSection &s = aSect[si];
					int16_t f = s.flags;
					const char *tstr = s.type > 0 && s.type < section_type_str ? section_type[s.type] : "error";
					if (f & (1 << ObjFileSection::OFS_MERGED)) {
						printf("Section %d: \"" STRREF_FMT "\": (Merged)",
							   reloc_idx, STRREF_ARG(PoolStr(s.name, str_orig)));
					} else if (f & (1 << ObjFileSection::OFS_DUMMY)) {
						if (f&(1 << ObjFileSection::OFS_FIXED)) {
							printf("Section %d: \"" STRREF_FMT "\": (Dummy [%s], fixed at $%04x)",
								   reloc_idx, STRREF_ARG(PoolStr(s.name, str_orig)), tstr, s.start_address);
						} else {
							printf("Section %d: \"" STRREF_FMT "\": (Dummy [%s], relative)",
								reloc_idx, STRREF_ARG(PoolStr(s.name, str_orig)), tstr);
						}
					} else {
						if (f&(1 << ObjFileSection::OFS_FIXED)) {
							printf("Section %d: \"" STRREF_FMT "\": (Fixed [%s] $%04x, $%x bytes)",
								reloc_idx, STRREF_ARG(PoolStr(s.name, str_orig)), tstr, s.start_address, s.end_address - s.start_address);
						} else {
							printf("Section %d: \"" STRREF_FMT "\": (Relative [%s], $%x bytes, align to $%x)",
								reloc_idx, STRREF_ARG(PoolStr(s.name, str_orig)), tstr, s.end_address - s.start_address, s.align_address);
						}
						strref export_append = PoolStr(s.exp_app, str_orig);
						if (export_append)
							printf("  Export as: \"" STRREF_FMT "\"", STRREF_ARG(export_append));
					}
					if (s.first_group >= 0)
						printf(" grouped with section %d", s.first_group);
					if (s.next_group >= 0)
						printf(" next in group: %d", s.next_group);
					if ((show & SHOW_CODE_RANGE) && s.output_size)
						printf(" code: $%x-$%x", (int)code_curr, (int)(code_curr + s.output_size));
					code_curr += s.output_size;
					printf("\n");
					reloc_idx++;
				}
			}
			
			if (show & SHOW_RELOCS) {
				int curRel = 0;
				for (int si = 0; si < hdr.sections; si++) {
					printf("section %d relocs: %d\n", si, aSect[si].relocs);
					for (int r = 0; r < aSect[si].relocs; r++) {
						struct ObjFileReloc &rs = aReloc[curRel++];
						printf("Reloc: section %d offset $%x base $%x bytes: %d shift: %d\n",
							   rs.target_section, rs.section_offset, rs.base_value, rs.bytes, rs.shift);
					}
				}
			}

			if (show & SHOW_MAP_SYMBOLS) {
				for (int mi = 0; mi < hdr.map_symbols; mi++) {
					struct ObjFileMapSymbol &m = aMapSyms[mi];
					printf("Symbol: \"" STRREF_FMT "\" section: %d value: $%04x%s\n",
						   STRREF_ARG(PoolStr(m.name, str_orig)), m.section, m.value, m.local ? " (local)" : "");
				}
			}

			if (show & SHOW_LABELS) {
				for (int li = 0; li < hdr.labels; li++) {
					struct ObjFileLabel &l = aLabels[li];
					strref name = PoolStr(l.name, str_orig);
					int16_t f = l.flags;
					int external = f & ObjFileLabel::OFL_XDEF;
					printf("Label: " STRREF_FMT " %s base $%x", STRREF_ARG(name), external==ObjFileLabel::OFL_XDEF ? "external" : "protected", l.value);
					if (external!=ObjFileLabel::OFL_XDEF)
						printf(" file %d", external);
					if (f & ObjFileLabel::OFL_CNST)
						printf(" constant");
					if (f & ObjFileLabel::OFL_ADDR)
						printf(" PC rel");
					if (f & ObjFileLabel::OFL_EVAL)
						printf(" evaluated");

					printf(" map index: %d\n", l.mapIndex);
				}
			}

			if (show & SHOW_LATE_EVAL) {
				for (int li = 0; li < hdr.late_evals; ++li) {
					struct ObjFileLateEval &le = aLateEval[li];
					strref name = PoolStr(le.label, str_orig);
					if (le.type == LEType::LET_LABEL) {
						printf("Late eval label: " STRREF_FMT " expression: " STRREF_FMT "\n",
							   STRREF_ARG(name), STRREF_ARG(PoolStr(le.expression, str_orig)));
					} else {
						printf("Late eval section: %d addr: $%04x scope: $%04x target: $%04x expression: " STRREF_FMT " %s\n",
							   le.section, le.address, le.scope, le.target, STRREF_ARG(PoolStr(le.expression, str_orig)),
							   late_type[le.type]);
					}
				}
			}

			if (show & SHOW_CODE_RANGE)
				printf("Code block: $%x - $%x (%d bytes)\n", (int)code_start, (int)size, (int)(size - hdr.bindata));

			// restore previous section
		} else
			printf("Not a valid x65 file\n");
	} else
		printf("Could not open %s\n", file);
}

int main(int argc, char **argv)
{
	const char *file = nullptr;
	uint32_t show =  0;
	for (int a = 1; a<argc; a++) {
		if (argv[a][0]=='-') {
			strref arg = argv[a]+1;
			if (arg.same_str("sections"))
				show = show | SHOW_SECTIONS;
			else if (arg.same_str("relocs"))
				show = show | SHOW_RELOCS;
			else if (arg.same_str("labels"))
				show = show | SHOW_LABELS;
			else if (arg.same_str("map"))
				show = show | SHOW_MAP_SYMBOLS;
			else if (arg.same_str("late_eval"))
				show = show | SHOW_LATE_EVAL;
			else if (arg.same_str("code"))
				show = show | SHOW_CODE_RANGE;
			else if (arg.same_str("all"))
				show = SHOW_SECTIONS | SHOW_RELOCS | SHOW_LABELS | SHOW_MAP_SYMBOLS | SHOW_LATE_EVAL | SHOW_CODE_RANGE;
		} else if (!file)
			file = argv[a];
	}

	if (!file) {
		printf("Usage:\ndump_x65 filename [-sections] [-relocs] [-labels] [-map] [-late_eval] [-code]\n");
		return 0;
	}

	if (argc>1) {
		ReadObjectFile(file, show ? show : SHOW_DEFAULT);
	}
}
